SDL based display emulation driver

vanwinkeljan opened the pull request:
This driver introduces an emulated LCD display for the native POSIX board.
The emulated display driver makes use of SDL2 to render the displays frame
buffer into a dedicated desktop window.

Bellow is a screenshot of the LVGL sample build and running for native_posix board:
![sdl_display_driver](https://user-images.githubusercontent.com/1586475/43304139-2ca1d3f6-9173-11e8-9555-d77a9667a2c6.png)

aescolar commented:
Relate dto #7833 & #6826

SebastianBoe commented on ext/lib/gui/Kconfig:
```diff
@@ -0,0 +1,11 @@
+
+menu "Graphical user interface"
+
+source "ext/lib/gui/lvgl/Kconfig"
```
rsource "lvgl/Kconfig"

is cleaner

SebastianBoe reviewed (changes requested):
From a build perspective this is OK.

Except that it disallows using cmake 3.8, but 3.8 is arguably the best version because it is not affected by bug: https://github.com/zephyrproject-rtos/zephyr/issues/8355

So to merge this one must either drop the cmake bump or keep the cmake bump and fix #8355 somehow.

Also, bumping the version requires an email to the mailing list that warns about the dependency change and an update of the getting started docs.

vanwinkeljan commented:
@SebastianBoe Note that only the two last commits are the actual pull request, the remaining commits are part of #6826 and #7833. I will take your comments along in #6826.

pizi-nordic commented on samples/gui/lvgl/src/main.c:
```diff
@@ -37,7 +43,7 @@ void main(void)
 	lv_obj_t *hello_world_label;
 	lv_obj_t *count_label;
 
-	display_dev = device_get_binding("ILI9340");
```
IMHO display API should be generic enough to make this line sole change in the sample.
(In this case changing SDL pixel format should be enough).

vanwinkeljan replied:
It is not clear here what you mean exactly:
1) take care of the conversion in both, SDL and ILI9340, display drivers
-> In this case best put discuss this in display driver API pull request (#7833)
2) Extend the kconfig options of the SDL display driver to select pixel format at compile time
-> This could lead to a lot of options in the config
3) Detect pixel format at run time for the SDL display
-> Possibly the best solution, but for simplicity only detect format on first write

pizi-nordic replied:
Ideally the user of display API should not change behaviour depending on display type. For example, the same code should correctly drive B&W and color display as long as both uses the same resolution. Of course performance and image quality will be compromised in such case, as some pixel format conversion will be needed. However it "should work" in my opinion.

I do not know on what level of display API such conversion/abstraction might implemented. Basing on this PR I assume that display.h just abstracts hardware and the LittlevGL provides high level primitives, thus the code use LittlevGL should be able to drive any display (especially if there will be API to fetch basic display properties like resolution and the sample will use this API in order to create correct picture).

vanwinkeljan replied:
Will be handled in PR #6826 by the GUI library

vanwinkeljan commented on cmake/app/boilerplate.cmake:
```diff
@@ -27,13 +27,11 @@ if(uname_output MATCHES "MSYS")
+# set to OLD to avoid copy-pasting cmake_minimum_required across application
 # CMakeLists.txt files.
-cmake_minimum_required(VERSION 3.8.2)
+cmake_minimum_required(VERSION 3.11)
```
it is actually part of another pending pull request (#6826), problem is that it is a dependency due to the sample that is used for testing

vanwinkeljan commented on include/display.h:
```diff
@@ -0,0 +1,156 @@
+ * @brief Callback API for writing bitmap
+ * See display_write_bitmap() for argument description
+ */
+typedef int (*display_write_bitmap_api)(
```
Moved comment to original pull request #7833

vanwinkeljan commented on include/display.h:
```diff
@@ -0,0 +1,156 @@
+ * @brief Callback API to check if pixel format is supported by display
+ * See display_is_pixel_format_supported() for argument description
+ */
+typedef bool (*display_is_pixel_format_supported_api)(
```
Moved comment to original pull request #7833

vanwinkeljan commented on include/display.h:
```diff
@@ -0,0 +1,156 @@
+ * @brief Display driver API
+ * API which a display driver should expose
+ */
+struct display_driver_api {
```
Moved comment to original pull request #7833

vanwinkeljan commented on drivers/display/display_ili9340.c:
```diff
@@ -92,50 +107,54 @@ int ili9340_init(struct device *dev)
 	struct ili9340_data *data = (struct ili9340_data *)dev->driver_data;
 
+	if (pixel_format != PIXEL_FORMAT_RGB_888) {
+		SYS_LOG_ERR("Pixel format not supported by ILI9340");
```
Moved comment to original pull request #7833

vanwinkeljan commented on ext/lib/gui/lvgl/lv_conf.h:
```diff
@@ -0,0 +1,321 @@
+#define LV_MEM_AUTO_DEFRAG  1               /*Automatically defrag on free*/
+#else       /*LV_MEM_CUSTOM*/
+#define LV_MEM_CUSTOM_INCLUDE <kernel.h>   /*Header for the dynamic memory function*/
+#define LV_MEM_CUSTOM_ALLOC   k_malloc       /*Wrapper to malloc*/
```
Moved comment to original pull request #6826

vanwinkeljan commented on ext/lib/gui/lvgl/lv_conf.h:
```diff
@@ -0,0 +1,321 @@
+ *=================*/
+
+/*Input device settings*/
+#define LV_INDEV_READ_PERIOD            50                     /*Input device read period in milliseconds*/
```
Moved comment to original pull request #6826

vanwinkeljan commented on ext/lib/gui/lvgl/lv_conf.h:
```diff
@@ -0,0 +1,321 @@
+
+/*Text settings*/
+#define LV_TXT_UTF8           CONFIG_LVGL_UTF_8  /*Enable UTF-8 coded Unicode character usage */
+#define LV_TXT_BREAK_CHARS    " ,.;:-_"          /*Can break texts on these chars*/
```
Moved comment to original pull request #6826

vanwinkeljan commented on ext/lib/gui/lvgl/lv_conf.h:
```diff
@@ -0,0 +1,321 @@
+/*Compiler settings*/
+#define LV_ATTRIBUTE_TICK_INC                 /* Define a custom attribute to `lv_tick_inc` function */
+#define LV_ATTRIBUTE_TASK_HANDLER             /* Define a custom attribute to `lv_task_handler` function */
+#define LV_COMPILER_VLA_SUPPORTED    1        /* 1: Variable length array is supported*/
```
Moved comment to original pull request #6826

vanwinkeljan commented on ext/lib/gui/lvgl/lv_conf.h:
```diff
@@ -0,0 +1,321 @@
+/* More info about fonts: https://littlevgl.com/basics#fonts
+ * To enable a built-in font use 1,2,4 or 8 values
+ * which will determine the bit-per-pixel */
+#define LV_FONT_DEFAULT        &lv_font_dejavu_20     /*Always set a default font from the built-in fonts*/
```
Moved comment to original pull request #6826

vanwinkeljan commented on samples/gui/lvgl/src/main.c:
```diff
@@ -0,0 +1,104 @@
+	lv_flush_ready();
+}
+
+void lvgl_tick_handler(struct k_timer *timer_id)
```
Moved comment to original pull request #6826

vanwinkeljan commented:
@aescolar Could review the last commit ([native_posix: Added polling for SDL events](https://github.com/zephyrproject-rtos/zephyr/pull/9149/commits/240cdfd941bd661d7179e9993e8eca18b149867c))?

aescolar commented on boards/posix/native_posix/CMakeLists.txt:
```diff
@@ -12,6 +12,15 @@ zephyr_library_sources(
 	)
 
+if(CONFIG_HAS_SDL)
+	find_package(PkgConfig REQUIRED)
```
For the cmake parts better if @SebastianBoe would take a look, but it looks a bit funny to have this code almost being a replica of the one for the driver.

aescolar commented on boards/posix/native_posix/hw_models_top.c:
```diff
@@ -125,6 +126,10 @@ void hwm_find_next_timer(void)
 {
 	while (1) {
+#ifdef CONFIG_HAS_SDL
+		hwm_sdl_poll_events();
```
Not goody. (if you place code here you have no control about how often it will be called)
If you want your code to be a HW model, and to be called periodically (say every 30 simulated ms), add a timer similar to `hw_timer_timer`:
* Add a new value in next_timer_index 
* Add your a pointer to your timer in Timer_list, and a call to you function in hwm_main_loop().
* Be sure to update your timer (`+=30e3`) into a future time value each time your function is called.

aescolar commented on drivers/display/Kconfig.sdl:
```diff
@@ -10,6 +10,7 @@ menuconfig SDL_DISPLAY
 	bool "SDL based emulated display"
 	depends on SOC_POSIX
```
Really you depend on NATIVE_POSIX (not SOC_POSIX) as the model is part of it.

aescolar commented on boards/posix/native_posix/sdl_model.c:
```diff
@@ -0,0 +1,61 @@
+			hwm_sdl_handle_window_event(&event);
+			break;
+		case SDL_QUIT:
+			raise(SIGTERM);
```
why do you raise SIGTERM instead of calling `posix_exit`?

vanwinkeljan replied:
Calling posix_exit would immediately exit the application and bypass the exit code in hwm_sleep_until_next_timer (currently only printing the simulated time)

aescolar replied:
* `posix_exit()` is the exit point for native_posix, which carries all necessary pre-exit steps. The "stopped at" print in `hwm_sleep_until_next_timer()` is just a silly nicety (linked with the --stop-at command line option). (thinking loud: Maybe I should just remove that print if it confuses the intent..)
* The signals were hooked to allow people to Ctrl+C/kill and still close in a controlled manner. They are not meant to be a normal way of closing. (Though they could be used from, say, a possible future asynchronous thread if there would be no better option).
* Having the signal flag exit thru the same place as the stop_at was just for pseudo code economy reason.

vanwinkeljan replied:
ok I will call posix_exit

aescolar commented on boards/posix/native_posix/sdl_model.c:
```diff
@@ -0,0 +1,61 @@
+/* Copyright (c) 2018 dXplore
```
Do you overall want to have this code as a HW model?
It feels to me it would more naturally fit as part of the driver.
If you need to run your initialization before Zephyr boots you can use `NATIVE_TASK` to register hwm_sdl_init to be called on `PREBOOT`.
You can create a Zephyr thread for the driver to update the window (or have a separate native pthread if you feel comfy enough)

vanwinkeljan replied:
Problem is that if it is made part of the display driver it depends on the driver to be present.
It could be the case that the SDL event loop is used to simulate other HW behavior such as input events (e.g. user input via keyboard/mouse).

Creating a thread is definitely not an option as SDL expects that the events are polled from the main thread.

aescolar replied:
(disclaimer I never played with SDL) In SDL can you not register a callback to be called when an event arrives instead of constantly polling for them?

> It could be the case that...

I guess one could imagine some other GUI based "HW models" for native_posix based on SDL. But as of now the only SDL based component is this proposed display driver. This split seems to be complicating things quite a bit. If possible I would propose the HAS_SDL code be made part of the driver by now, and if at some point in the future there is some other driver/model which also needs some of the SDL handling code, then that be moved to a common place.

vanwinkeljan replied:
Yes there is callback support in SDL but to my surprise it still requires the poll loop (to be more precisely a periodic call to [SDL_PumpEvents](https://wiki.libsdl.org/SDL_PumpEvents)).

Actually this code came to live when I was looking into using SDL to capture keyboard events to interact with the GUI. As this code was already useful for this PR I just appended it to the commit list.

aescolar replied:
So, a separate pthread (asynchronous to the other native_posix threads) dedicated to SDL, which would do the SDL initialization, and after, would have an infinite loop blocking on `SDL_WaitEvent()` would not be a possibility?

If you really need to have the polling code in the 1st thread, then, please do the timer thingy I mentioned above.

vanwinkeljan replied:
Thx for all the input I will go for the timer thingy especially because of the warning in the SDL_PumpEvents API documentation: 
> WARNING: This should only be run in the thread that initialized the video subsystem, and for extra safety, you should consider only doing those things on the main thread in any case.

vanwinkeljan replied:
@aescolar Tried a separate pthread and it seems to work

aescolar commented on drivers/display/Kconfig.sdl:
```diff
@@ -0,0 +1,45 @@
+	bool "SDL based emulated display"
+	depends on SOC_POSIX
+	default n
+	select HAS_SDL
```
The farther I go thru the code the mode I feel the "`HAS_SDL` HW model" is really a piece of the driver.

aescolar commented on drivers/display/Kconfig.sdl:
```diff
@@ -0,0 +1,45 @@
+
+menuconfig SDL_DISPLAY
+	bool "SDL based emulated display"
+	depends on SOC_POSIX
```
ARCH_POSIX would be enough (if you move the "HW model" into the driver)

aescolar replied:
You do depend on NATIVE_POSIX, not just SOC_POSIX, or?

aescolar commented on drivers/display/Kconfig.sdl:
```diff
@@ -0,0 +1,45 @@
+menuconfig SDL_DISPLAY
+	bool "SDL based emulated display"
+	depends on SOC_POSIX
+	default n
```
I guess you could set this driver as default one `if DISPLAY` for native posix in https://github.com/zephyrproject-rtos/zephyr/blob/master/boards/posix/native_posix/Kconfig.defconfig

aescolar commented on drivers/display/display_sdl.h:
```diff
@@ -0,0 +1,8 @@
+/* Copyright (c) 2018 dXplore
```
2 comments:
* for consistency you may want to lower the copyright message one line
* is "dXplore" an actual legal entity?

vanwinkeljan replied:
on the second point: no not yet

aescolar commented on drivers/display/Kconfig.sdl:
```diff
@@ -0,0 +1,45 @@
+	default n
+	select HAS_SDL
+	help
+	Enable SDL based emulated display compliant with display driver API.
```
You may want to indent the help messages (also below)

aescolar commented on boards/posix/native_posix/hw_models_top.c:
```diff
@@ -92,12 +98,17 @@ static void hwm_sleep_until_next_timer(void)
 	}
 
-	if (signaled_end || (simu_time > end_of_time)) {
+	_SAFE_CALL(pthread_mutex_lock(&exit_mutex));
```
Technically what you do is correct: you lock a mutex to protect access to the shared variables `hwm_exit_request` and `hwm_exit_code`.
But this loop executes for each simulation "delta cycle". Locking and unlocking a mutex in each iteration adds quite a bit of ovehead.
So, instead of doing it this way, I would:
* remove `hwm_exit_request` and have `hwm_main_loop_exit` set `signaled_end`.
* Have the mutex here protect only the access to hwm_exit_code (so move the mutex lock call inside the if)

Why: signaled_end starts as 0, and is only set to something != 0 by another thread when it wants things to end. It is safe for the hwm thread to read that variable at any moment without protection, because in the worst case it just waits for one hw model cycle more before stopping. Here, there is no race we need to protect against. 
(see other comments for more thoughts)

aescolar commented on boards/posix/native_posix/hw_models_top.c:
```diff
@@ -92,12 +98,17 @@ static void hwm_sleep_until_next_timer(void)
-	if (signaled_end || (simu_time > end_of_time)) {
+	_SAFE_CALL(pthread_mutex_lock(&exit_mutex));
+	if (signaled_end || (simu_time > end_of_time) || hwm_exit_request) {
+		int exit_code = hwm_exit_request ? hwm_exit_code : 0;
```
Minor: `hwm_exit_code` is 0 unless it was set differently in `hwm_main_loop_exit` (with `hwm_exit_request`). So you may as well have `int exit_code = hwm_exit_code;` (without the if), and add a comment instead if you feel its needed.

aescolar commented on boards/posix/native_posix/hw_models_top.c:
```diff
@@ -92,12 +98,17 @@ static void hwm_sleep_until_next_timer(void)
-		posix_exit(0);
+		posix_exit(exit_code);
 	}
+	_SAFE_CALL(pthread_mutex_unlock(&exit_mutex));
```
This one is not needed if you only want to protect the hwm_exit code, and just have the mutex_lock call inside the if.

aescolar commented on boards/posix/native_posix/hw_models_top.c:
```diff
@@ -190,4 +201,29 @@ void hwm_cleanup(void)
+ *
+ */
+void hwm_main_loop_exit(int exit_code)
+{
```
Just a silly comment related to the previous comments: if you did not have the `exit_code` all this function body could be replaced with just `signaled_end = 1`.

vanwinkeljan replied:
correct but in this way the thread can signal that an exit was performed due to a failure and the error can be propagated to the command line

aescolar commented on boards/posix/native_posix/sdl_event_loop.c:
```diff
@@ -0,0 +1,89 @@
+	_SAFE_CALL(pthread_mutex_unlock(&mutex));
+}
+
+NATIVE_TASK(create_sdl_event_loop_thread, PRE_BOOT_1, 1);
```
Minor: You may prefer `PRE_BOOT_2`. It is relatively likely that the user passes some invalid command line arguments to the program, and it exists before reaching `PRE_BOOT_2`, while you don't need to start your SDL thread before that.

aescolar commented on ext/lib/gui/lvgl/lvgl_mem_kernel.c:
```diff
@@ -0,0 +1,25 @@
+/*
```
Separate issue: Quite a few of this files placed in ext/, would seem to not be copied directly from an external project, but be actual written by you directly for integrating it with Zephyr?

vanwinkeljan replied:
Correct they are part of a pending PR #6826 and this PR sample code depends on it.

aescolar reviewed (commented):
I think it is much better now. this should lead to less overhead, more responsiveness on the GUI part, and better decoupling of future code.

vanwinkeljan committed:
driver: SDL based display emulation driver

This driver introduces an emulated LCD display for the native POSIX
board. The emulated display driver makes use of SDL2 to render the
displays frame buffer into a dedicated desktop window.

Signed-off-by: Jan Van Winkel <jan.van_winkel@dxplore.eu>

vanwinkeljan committed:
samples: Added native posix board to cfb sample

Added support for native posix board to character framebuffer sample.

Signed-off-by: Jan Van Winkel <jan.van_winkel@dxplore.eu>

vanwinkeljan commented:
Changes:
* Removed dependency on lvgl (#6826) and use character framebuffer subsystem instead (including sample).
  @nashif can you remove DNM label as this PR has no longer any dependencies on other PR?
* Added extra posix signal to sdl event loop to prevent crash in case events pop up during driver initialization
* Small corrections in cfb and SSD1306 & SDD1673 drivers (#9035)
  @jfischer-phytec-iot  please review

aescolar reviewed (changes requested):
Nothing too serious. Note that I only focused on the native_posix related parts. I have not reviewed the driver side.

aescolar reviewed (dismissed):
Nothing too serious. Note that I only focused on the native_posix related parts. I have not reviewed the driver side.

aescolar commented:
It seems the underlying GitHub entry holding this PR got quite corrupted. I see my last review either not at all, or several times when I refresh this page. Maybe you want to close the PR and create a new one with this same branch?

vanwinkeljan commented:
@aescolar Github is experiencing some difficulties (https://status.github.com/messages) at the moment let see how it turns out else I will create a new PR

vanwinkeljan commented:
recheck

aescolar commented:
As this PR is adding a new "peripheral" for native_posix, it would be nice adding a new section in https://docs.zephyrproject.org/latest/boards/posix/native_posix/doc/board.html#peripherals for it. You could also mention there that it will require `pkg-config` and `libsdl2-dev:i386`.
At least in my case, I also needed to `export PKG_CONFIG_PATH=/usr/lib/i386-linux-gnu/pkgconfig/` before calling cmake to get it to build the demo app.

aescolar commented:
@vanwinkeljan I have just tried the app in a VM, and it runs very slowly. Are you experiencing that in your side?

vanwinkeljan commented:
@aescolar In my case the scrolling line wraps around in 1 to 2 seconds.
In case it is much slower at your side could this be caused by the fact that the main loop is a tight loop (`k_cpu_idle` has been added)?

aescolar commented:
> In case it is much slower at your side

It is around 1 second here.. I guess "slow" can have different meanings :)

aescolar commented:
@SebastianBoe : Is your review still valid?

aescolar commented:
@vanwinkeljan : I have just run that sample app with valgrind's callgrind tool. In 1 scroll, there seems to be 450.000 calls to `SDL_UpdateTexture()` which lead to a baffling amount of activity inside the SDL library (over 2.5 thousand million == 2.5 short billion instructions in the SDL library are executed for each wrap).

vanwinkeljan commented:
@aescolar  I'm not surprised about the number 450.000. This s due that SDL rendering buffer is ARGB and the CFB sample uses monochrome displays, the function ```sdl_display_write_mono``` naively takes care of the conversion pixel per pixel. Further CFB always writes the complete screen and the screen size is 250x120 so it seems per wrap there are 15 full screen writes.

On the other hand I'm a bit surprised by the number of calls in the SDL library I was assuming that the reddening was delayed until a single call in the function ```sdl_display_write```.

vanwinkeljan commented:
@aescolar Quickly googled it and as far as I understand it will for each call to ```SDL_UpdateTexture``` transfer the data to the GPU and for the monochrome display this is per pixel.  I could try to optimize it a bit but no idea how much it will improve, first time playing around with SDL.

Fyi from the remarks section of ```SDL_UpdateTexture``` (https://wiki.libsdl.org/SDL_UpdateTexture ;)
> This is a fairly slow function, intended for use with static textures that do not change often.
>
>If the texture is intended to be updated often, it is preferred to create the texture as streaming and use the locking functions referenced below. While this function will work with streaming textures, for optimization reasons you may not get the pixels back if you lock the texture afterward.

aescolar commented:
> I could try to optimize it a bit but no idea how much it will improve

I would guess that if you built the whole display in one rectangle, and called UpdateTexture once per sdl_display_write, it would speed up things 30.000 times.

vanwinkeljan commented:
>I would guess that if you built the whole display in one rectangle, and called UpdateTexture once per sdl_display_write, it would speed up things 30.000 times.

That was actually my initial plan but that would have require an extra buffer and I had already the texture available. I will try the streaming texture + locking as mentioned above and see how it turns out, if it does not improve I will create local buffer.

aescolar commented:
> that would have require an extra buffer

That extra buffer is trivially cheap in native_posix. You could have it preallocated since you initialized the display driver.

vanwinkeljan commented:
* Added display peripheral to documentation
* Performance improvement: Use ```SDL_LockTexture```/```SDL_UnlockTexture``` instead of ```SDL_UpdateTexture```

aescolar commented:
@vanwinkeljan : Uhm..., trying this new version I'm meeting with two errors occasionally. Sometimes everything runs fine, sometimes I get a:
```
XIO:  fatal IO error 14 (Bad address) on X server ":0.0"
      after 288 requests (286 known processed) with 0 events remaining.
``` 
Sometimes (more often) I get a segfault in the SDL library.
Both always after the `cfb_framebuffer_finalize()` finalize has completed and before the next iteration of the loop starts. So they seem related.

aescolar commented on drivers/display/display_sdl.c:
```diff
@@ -0,0 +1,323 @@
+			pixel |= *(byte_ptr + 1) << 8;
+			pixel |= *(byte_ptr + 2);
+			*((u32_t *)disp_buf) = pixel;
+			*disp_buf += 4;
```
do you want to add 4 to the data? or move the pointer 4 bytes ahead?

vanwinkeljan replied:
nice catch

aescolar commented:
@vanwinkeljan By the way, regarding the performance issue, it is obviously better. But you can also see the numbers here (both for 1 full scroll). Old commit 6afa8a53a469661f:
![image](https://user-images.githubusercontent.com/32480799/47613990-bf0d3c80-da98-11e8-8dbd-7450fedf1a9e.png)
New commit (b389754472e4378f699):
![image](https://user-images.githubusercontent.com/32480799/47613992-c7fe0e00-da98-11e8-8ed9-f05b70da797b.png)
The 3 first columns are: 
* Total instructions executed in that function (including children)
* Total instructions executed in the function itself (no children)
* Number of times the function was called

The total zephyr.exe instruction count in the new commit was 717M

aescolar reviewed (changes requested):
Occasional crashes

vanwinkeljan commented:
@aescolar Thx for all the debugging possibly this is caused by the fact that SDL assumes everything is rendered from a single thread.

As mentioned in a earlier comment:
>Thx for all the input I will go for the timer thingy especially because of the warning in the SDL_PumpEvents API documentation:
>
>WARNING: This should only be run in the thread that initialized the video subsystem, and for extra safety, you should consider only doing those things on the main thread in any case.

So two options here:
1) Move everything towards the main hw loop
2) Put a message queue in place between the event loop thread and the main hw loop.
I already had a proposal ready to support keyboard based GPIO via SDL that contains this solution.
You could have a look at following 2 difss to get an idea:
https://github.com/vanwinkeljan/zephyr/commit/c762321b84ead7c1951c96eb3b8b52765f95096a#diff-b34741663195be2a7c40e7fa916080f0
https://github.com/vanwinkeljan/zephyr/commit/c762321b84ead7c1951c96eb3b8b52765f95096a#diff-d5441623e86f5884cfbfaa5e58f58daf

vanwinkeljan commented:
@aescolar I will have a closer look at buffer write functions. In the end I managed to reproduce the issue by putting ```k_cpu_idle``` back in place and sending the window to the background and then when I bring the window to the foreground the crash occurs. But I still have the feeling that it is just a multi threading issue, especially as the issue is gone after placing a ```return``` at the beginning of the function ```sdl_handle_window_event```.

aescolar commented:
> In the end I managed to reproduce the issue ..

Curious, here it crashes maybe one out of every 3 times on its own (no need to touch anything)

> But I still have the feeling that it is just a multi threading ..

It may be. But note that the thread that calls the sdl_write function is anyhow a Zephyr thread, while the HW models are another thread.
( It can be quite annoying to track down a bug like this which is exposing itself deep in the guts of a 3rd party library. )

vanwinkeljan commented:
@aescolar Added synchronous event handling code to handle the render event in the hw loop.
This should prevent concurrent calls to ```SDL_RenderPresent``` as the zephyr threads are not executing at the same time as the hw main loop thread. Could you test this from your side to see if it resolves the problems, for me it is at least solved.

nashif commented:
just coming back to trying this out after a while...

i am getting this:
```
[100%] Built target zephyr_prebuilt
Scanning dependencies of target run
dbus[1429]: arguments to dbus_message_new_method_call() were incorrect, assertion "path != NULL" failed in file ../../dbus/dbus-message.c line 1362.
This is normally a bug in some application using the D-Bus library.

```

Any hints?

vanwinkeljan commented:
@nashif 
> Any Hints?

Are you on ubuntu 18.04?
If this is the case see: https://bugs.launchpad.net/ubuntu/+source/libsdl2/+bug/1775067 
Comment 7 has a posible workaround.

nashif commented:
> Are you on ubuntu 18.04?

Fedora 28

vanwinkeljan commented:
> > Are you on ubuntu 18.04?
> 
> Fedora 28

Does the workaround work?

nashif commented:
> Does the workaround work?

yeah, workaround 1 in comment 7 works

vanwinkeljan commented:
@aescolar any chance you found some time to test the latest fixes?
@nashif Did you encounter any instabilities during your testing?
@SebastianBoe could you revisit you review?

aescolar commented on boards/posix/native_posix/hw_models_top.c:
```diff
@@ -134,13 +152,18 @@ void hwm_main_loop(void)
 			/* LCOV_EXCL_STOP */
 		}
+#ifdef CONFIG_HAS_SDL
+		sdl_handle_synchronious_events();
```
Please do not set code to execute in each delta cycle. You do not need to and it would bog down everything else.

vanwinkeljan replied:
I could move it into the case, but my idea here was that you would like to act as fast as possible on the asynchronous events and as such check it for every expired timer.

aescolar commented on boards/posix/native_posix/hw_models_top.c:
```diff
@@ -134,13 +152,18 @@ void hwm_main_loop(void)
 			hw_irq_ctrl_timer_triggered();
 			break;
+		case SDLEVENTTIMER:
+			break;
```
If you create a new timer for the SDL library you should call the function to runs when the timer triggers from here. An empty case is not useful. The function you call from here should then update the timer trigger to some future value (or `NEVER`).

vanwinkeljan replied:
Actually ```sdl_handle_synchronious_events``` is updating the timer and is putting it 10ms into the future after an event occurred

aescolar commented on boards/posix/native_posix/hw_models_top.c:
```diff
@@ -26,20 +30,28 @@ static u64_t end_of_time = NEVER; /* When will this device stop */
 extern u64_t irq_ctrl_timer;
+#ifdef CONFIG_HAS_SDL
+extern u64_t sdl_event_timer;
+#else
```
Better instead ifdef the code below that is only used for SDL to not affect other use cases.

aescolar commented on boards/posix/native_posix/hw_models_top.c:
```diff
@@ -190,4 +213,30 @@ void hwm_cleanup(void)
 }
 
+/**
+ * Send a request to the HW model main loop to perform a exit.
```
perform a+n+ exit.

aescolar commented:
> any chance you found some time to test the latest fixes?

@vanwinkeljan : I just took a quick look now.. I hope I have some time in the next couple of days.

vanwinkeljan commented:
@aescolar at first thx for the quick feedback

> I just took a quick look now.. But I hope I'll have some time in the next couple of days.

Actually I was only hooping for a test, I just grapt this code from some "_work in progress_" for test purposes.

aescolar commented:
>  I just grapt this code from some "work in progress" for test purposes

Then if you don't mind let's add the DNM tag back.

vanwinkeljan commented:
@aescolar In case the crashes do not show up in your test I will rework the code to handle all events in the ```SDLEVENTSTIMER``` case of the ```hwm_main_loop```

aescolar commented:
> In case the crashes do not show up in your test

They did not.
But note that all I do is run it in a Xubuntu18.04 VM , so you do not need me just for testing it a bit..

aescolar commented:
I wonder if the reason why it does not crash now is because of changing the threading, or because of the changes you did in the driver itself with the texture locking and so forth.

vanwinkeljan commented:
> But note that all I do is run it in a Xubuntu18.04 VM , so you do not need me just for testing it a bit..

I did test locally first but still thought it was a good idea it was tested on different machines.

> I wonder if the reason why it does not crash now is because of changing the threading, or because of the changes you did in the driver itself with the texture locking and so forth.

Once I found a scenario to reproduce the issue was still there with the reworked driver  and it always originated from ```SDL_RenderPresent```, stack trace indicated that came from inside the OpenGL driver. Making sure that the ```SDL_RenderPresent``` was called sequentially  resolved the issue.

vanwinkeljan commented:
Updated code to handle SDL events in ```hwm_main_loop``` instead of dedicated event loop tot prevent concurrent calls to SDL functions.

vanwinkeljan commented:
Moved SSD1673 & CFB commits to separate PR #11351, only 3 last commits are now part of this PR

aescolar commented on boards/posix/native_posix/sdl_events.h:
```diff
@@ -0,0 +1,14 @@
+ * SPDX-License-Identifier: Apache-2.0
+ */
+#ifndef ZEPHYR_BOARD_POSIX_NATIVE_POSIX_SDL_EVENTS_H
+#define ZEPHYR_BOARD_POSIX_NATIVE_POSIX_SDL_SDL_EVENTS_H
```
Not matching guard and ifdef

vanwinkeljan replied:
nice catch

aescolar commented:
> Moved SSD1673 & CFB commits to separate PR

I'm a bit confused with this: the SSD1673  commits are still in this PR branch, did you forget to remove them?

aescolar commented:
@vanwinkeljan : shouldn't the commits:
`driver: SDL based display emulation driver`
`native_posix: Use synchronous SDL event handling`
be squashed toghether?

aescolar commented on boards/posix/native_posix/doc/board.rst:
```diff
@@ -519,6 +519,15 @@ The following peripherals are currently provided with this board:
+  render the display content. Underlying the display driver is using the
+  `SDL2`_ library for rendering and as such requires a 32-bit version of the
+  SDL2 library installed on the host machine. Further there is a dependency on
+  pkg-config to correctly pickup the SDL2 install path and compiler flags.
```
I feel this description alone is going to be too little for quite a few users. (Even when we give exact steps with packages to install some people has issues..)
In Ubuntu 18.04 I needed to install `pkg-config` and `libsdl2-dev:i386`, but also I needed to `export PKG_CONFIG_PATH=/usr/lib/i386-linux-gnu/pkgconfig/` before calling cmake. This is not too trivial and probably will cause a bug report as soon as somebody tries it.

vanwinkeljan replied:
Added a section on which packages to install and the export

aescolar commented on drivers/display/display_sdl.c:
```diff
@@ -0,0 +1,311 @@
+	}
+
+	disp_data->renderer =
+	    SDL_CreateRenderer(disp_data->window, -1, SDL_RENDERER_ACCELERATED);
```
Note that you are never cleaning up on exit.
You can run it with `valgrind --leak-check=full zephyr/zephyr.exe` and will see there is ~5MB left allocated.
(At least I prefer not having (too much) left over to be sure I can still see real issues with memory leaks)

vanwinkeljan replied:
Added exit functions ```sdl_display_cleanup``` and ```sdl_cleanup```, first function is cleaning up window, renderer and texture; the second function handles the SDL exit.

Even after calling all this cleanup functions still ~3MB is in use, some googling reveals this mainly due to X11 en OpenGL implementation.

aescolar replied:
> Even after calling all this cleanup functions still ~3MB is in use,

Oh well.. I guess the SDL library is designed with a different mentality.

aescolar replied:
That would be nice

aescolar reviewed (dismissed):
Please see the few spread comments. It seems this is getting closer and closer to be ready.
I did not experience any crash while playing with it this time

vanwinkeljan commented:
> > Moved SSD1673 & CFB commits to separate PR
> 
> I'm a bit confused with this: the SSD1673 commits are still in this PR branch, did you forget to remove them?

This is a bit of a github issue as it can not track dependencies between PRs.

What happend here is that this PR now dependence on PR #11351 so it shows also all commits part PR #11351

vanwinkeljan commented:
> @vanwinkeljan : shouldn't the commits:
> `driver: SDL based display emulation driver`
> `native_posix: Use synchronous SDL event handling`
> be squashed together? (if you want to go with the synchronous solution, that is. But if you are not sure yet then leave it in 2 commits)

Yes they should but I thought it would be more clear to reviewers to see what got changed.

aescolar commented:
> This is a bit of a github issue

This PR is a request to merge the branch `sdl_display` into `master`. 
Now, the last commit `sdl_display` has in common with `master` is 644b31d2a4245104ae. Everything after that is what this pull request is requesting to merge (and would be merged if somebody pressed the merge button).
If that set of commits has something in common with other branches, or other PRs, is not relevant from GitHub's point of view (One cannot expect GitHub to read the author's mind and guess what is it that the author is trying to do with common commits in different branches/PRs/repos).

aescolar commented on boards/posix/native_posix/doc/board.rst:
```diff
@@ -519,6 +519,20 @@ The following peripherals are currently provided with this board:
+  SDL2 library installed on the host machine. Further there is a dependency on
+  pkg-config to correctly pickup the SDL2 install path and compiler flags.
+
+  On a Ubuntu 18.04 host system this results in the needs of installing the
```
`On a Ubuntu 18.04 host system this results in the needs of installing the` seems a too convoluted. Maybe just "In Ubuntu 18.04 install the xx and yy packages,"? instead. 
But better if @dbkinder proposes.

vanwinkeljan commented:
> This is a bit of a github issue

@aescolar Maybe I had better word this as it is a "highly requested feature", of course it would be up to the author to mark the dependencies.

aescolar commented:
@vanwinkeljan : Is this PR actually blocked by #11351? or simply this PR branch branched from that other one?

vanwinkeljan commented:
@aescolar Thx for all the review effort!

dbkinder commented on boards/posix/native_posix/doc/board.rst:
```diff
@@ -519,6 +519,20 @@ The following peripherals are currently provided with this board:
   Bluetooth specification version 4.0 or greater).
 
+**Display driver**:
+  A display driver is provided that will create a window on the host machine to
```
How about:
```
**Display driver**:
  A character frame buffer (CFB) display driver is provided that
  creates a window on the host machine to render display content. 
  (Currently, the CFB framework is limited to monochrome
  vertical tiled displays such as found on the :ref:`reel_board``.)

  This driver requires a 32-bit version of  the`SDL2`_ library
  on the host machine and ``pkg-config`` settings to correctly
  pickup the SDL2 install path and compiler flags.

  On a Ubuntu 18.04 host system, for example,  install the
  ``pkg-config`` and ``libsdl2-dev:i386`` packages, and configure
  the pkg-config search path with these commands::

     $ sudo apt-get install pkg-config libsdl2-dev:i386
     $ export PKG_CONFIG_PATH=/usr/lib/i386-linux-gnu/pkgconfig

```

Also, the picture you have in the commit message would be a nice to show here.  (I see though that you used a proportional font in that display.  Does the CFB driver handle proportional fonts or are only mono-spaced fonts supported?)

vanwinkeljan replied:
Thx for the proposal looks much better, will update the docs.

> Also, the picture you have in the commit message would be a nice to show here. (I see though that you used a proportional font in that display. Does the CFB driver handle proportional fonts or are only mono-spaced fonts supported?)

This picture that is shown is actually for the lvgl sample application which is part of PR #6826.
Once CFB came available with the introduction of the reel board it looked like a better idea to reuse that sample.

But I could make a screenshot of CFB in action but do you have any idea on the policy of putting binary files in to the repo?

dbkinder commented on samples/display/cfb/README.rst:
```diff
@@ -0,0 +1,35 @@
+Overview
+********
+
+This sample will display a vertical scrolling text string making use of the
```
To me, "vertical scrolling text string" means it would look like this:

    V
    E
    R
    T
    I
    C
    A
    L

I think this sentence should instead be:

    This sample displays character strings using the Character Frame Buffer
    (CFB) subsystem framework.

I was hoping to find documentation for this CFB subsystem, but alas I can't find it.  Sigh.

vanwinkeljan replied:
Point taken on the "vertical"

> I was hoping to find documentation for this CFB subsystem, but alas I can't find it. Sigh.

No there is not much of documentation on CFB but I did take some effort on documenting the sample in PR #11351 (which is till waiting for a review on the doc ;) )

dbkinder commented on samples/display/cfb/README.rst:
```diff
@@ -0,0 +1,35 @@
+Requirements
+************
+
+This sample requires a board and a display, the display needs to be supported
```
How about:

    This sample requires a supported board and CFB-supporting
    display, such as the :ref:`reel_board`.

dbkinder reviewed (approved):
Docs LGTM, thanks.

vanwinkeljan commented:
@SebastianBoe could you revisit you review?

SebastianBoe commented:
> @SebastianBoe could you revisit you review?

I see that the original review comment stated that this would require a bump in the CMake version. Which has not happened. Is this still valid then?

vanwinkeljan commented:
Correct there is no need to up step CMake anymore as _west_ will be used instead and actually the up step was not needed for this PR but for PR #6826.

nashif reviewed (approved):
please replace subsys: in your commit messages and put the actual subsys you are changing.

aescolar commented:
@nashif : I did. This PR is DNM, being blocked by #11351

vanwinkeljan commented:
@aescolar Thx was just typing my self that this PR blocked

vanwinkeljan commented:
@nashif replaced subsys with cfb in commit message

nashif commented:
please rebase

vanwinkeljan commented:
@nashif done
